File: src/OpStream.Client.Transports.gRPC/gRPCOpStreamClient.cs
using Grpc.Net.Client;
using Grpc.Core;
using OpStream.Shared.Messages.gRPC;
using OpStream.Shared.Messages;
using System.Text.Json;
using System.Collections.Concurrent;
using OpStream.Constants;
using Microsoft.Extensions.Options;

namespace OpStream.Client.Transports.gRPC;

/// <summary>
/// A gRPC-based implementation of the OpStream client transport.
/// Handles bidirectional streaming and request-response patterns over a single stream.
/// </summary>
public class gRPCOpStreamClient : IOpStreamClient
{
    private readonly string _peerId = Guid.NewGuid().ToString();
    private readonly GrpcChannel _channel;
    private readonly OpStreamService.OpStreamServiceClient _client;
    private readonly OpStreamCommentsService.OpStreamCommentsServiceClient _commentsClient;
    private readonly AsyncDuplexStreamingCall<ClientMessage, ServerMessage> _call;
    private readonly ConcurrentDictionary<string, TaskCompletionSource<ServerMessage>> _pendingRequests = new();
    private readonly Task _listenTask;
    private readonly CancellationTokenSource _cts = new();
    private Task? _commentListenTask;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        PropertyNameCaseInsensitive = true
    };

    /// <inheritdoc/>
    public event Func<ReadOnlyMemory<byte>, long, Task>? OnReceiveOp;

    /// <inheritdoc/>
    public event Action<Exception?>? OnDisconnected;

    /// <inheritdoc/>
    public event Func<IEnumerable<AwarenessState>, Task>? OnReceiveAwareness;

    /// <inheritdoc/>
    public event Action<string>? OnPeerDisconnected;

    // ─── Comment events ───────────────────────────────────────────────────────

    private event Func<CommentDto, Task>? _onCommentCreated;
    private event Func<CommentDto, Task>? _onCommentUpdated;
    private event Func<CommentDeletedDto, Task>? _onCommentDeleted;

    /// <inheritdoc/>
    public event Func<CommentDto, Task>? OnCommentCreated
    {
        add    => _onCommentCreated += value;
        remove => _onCommentCreated -= value;
    }

    /// <inheritdoc/>
    public event Func<CommentDto, Task>? OnCommentUpdated
    {
        add    => _onCommentUpdated += value;
        remove => _onCommentUpdated -= value;
    }

    /// <inheritdoc/>
    public event Func<CommentDeletedDto, Task>? OnCommentDeleted
    {
        add    => _onCommentDeleted += value;
        remove => _onCommentDeleted -= value;
    }

    /// <summary>
    /// Initializes a new instance of the gRPCOpStreamClient.
    /// </summary>
    /// <param name="options">Configuration options containing the server address.</param>
    public gRPCOpStreamClient(IOptions<OpStreamgRPCOptions> options)
    {
        _channel = GrpcChannel.ForAddress(options.Value.ServerAddress);
        _client = new OpStreamService.OpStreamServiceClient(_channel);
        _commentsClient = new OpStreamCommentsService.OpStreamCommentsServiceClient(_channel);
        _call = _client.Connect();
        _listenTask = ListenLoop();
    }

    /// <summary>
    /// Background task that listens for incoming messages from the server.
    /// </summary>
    private async Task ListenLoop()
    {
        try
        {
            await foreach (var message in _call.ResponseStream.ReadAllAsync(_cts.Token))
            {
                if (!string.IsNullOrEmpty(message.CorrelationId) && _pendingRequests.TryRemove(message.CorrelationId, out var tcs))
                {
                    tcs.SetResult(message);
                    continue;
                }

                // Handle events
                switch (message.MessageTypeCase)
                {
                    case ServerMessage.MessageTypeOneofCase.ReceiveOpEvent:
                        if (OnReceiveOp != null)
                        {
                            await OnReceiveOp.Invoke(message.ReceiveOpEvent.Payload.ToByteArray(), message.ReceiveOpEvent.NewRevision);
                        }
                        break;
                    case ServerMessage.MessageTypeOneofCase.ReceiveAwarenessEvent:
                        if (OnReceiveAwareness != null)
                        {
                            await OnReceiveAwareness.Invoke(message.ReceiveAwarenessEvent.Awareness.Select(FromProto));
                        }
                        break;
                    case ServerMessage.MessageTypeOneofCase.PeerDisconnectedEvent:
                        OnPeerDisconnected?.Invoke(message.PeerDisconnectedEvent.PeerId);
                        break;
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception ex)
        {
            OnDisconnected?.Invoke(ex);
        }
    }

    /// <inheritdoc/>
    public async Task<ClientJoinResult> ConnectAndJoinAsync(string documentId, string documentType, CancellationToken ct = default)
    {
        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<ServerMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pendingRequests[correlationId] = tcs;

        await _call.RequestStream.WriteAsync(new ClientMessage
        {
            CorrelationId = correlationId,
            JoinRequest = new JoinRequest
            {
                DocumentId = documentId,
                DocumentType = documentType,
                ClientProtoVersion = 1
            }
        });

        using var registration = ct.Register(() => tcs.TrySetCanceled());
        var response = await tcs.Task;
        
        if (response.MessageTypeCase != ServerMessage.MessageTypeOneofCase.JoinResponse)
        {
            throw new Exception("Unexpected response during Join");
        }

        var jr = response.JoinResponse;
        _commentListenTask = ListenCommentsLoop(documentId, _cts.Token);
        return new ClientJoinResult(jr.Revision, jr.Snapshot.ToByteArray(), jr.Awareness.Select(FromProto));
    }

    /// <inheritdoc/>
    public async Task<ClientOpResult> SendOpAsync(string documentId, ReadOnlyMemory<byte> payload, long baseRevision, CancellationToken ct = default)
    {
        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<ServerMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pendingRequests[correlationId] = tcs;

        await _call.RequestStream.WriteAsync(new ClientMessage
        {
            CorrelationId = correlationId,
            OpRequest = new OpRequest
            {
                DocumentId = documentId,
                Payload = Google.Protobuf.ByteString.CopyFrom(payload.Span),
                BaseRevision = baseRevision
            }
        });

        using var registration = ct.Register(() => tcs.TrySetCanceled());
        var response = await tcs.Task;

        if (response.MessageTypeCase != ServerMessage.MessageTypeOneofCase.OpResponse)
        {
            throw new Exception("Unexpected response during SendOp");
        }

        var or = response.OpResponse;
        return new ClientOpResult(or.Success, or.NewRevision, or.ErrorMessage);
    }

    /// <inheritdoc/>
    public async Task SendAwarenessAsync(string documentId, JsonElement data, CancellationToken ct = default)
    {
        await _call.RequestStream.WriteAsync(new ClientMessage
        {
            CorrelationId = "", // No correlation needed for one-way messages
            AwarenessRequest = new AwarenessRequest
            {
                DocumentId = documentId,
                DataJson = data.GetRawText()
            }
        });
    }

    // ─── Comment subscription ─────────────────────────────────────────────────

    private async Task ListenCommentsLoop(string documentId, CancellationToken ct)
    {
        try
        {
            using var call = _commentsClient.SubscribeComments(
                new SubscribeCommentsRequest { DocumentId = documentId },
                cancellationToken: ct);

            await foreach (var evt in call.ResponseStream.ReadAllAsync(ct))
            {
                switch (evt.EventTypeCase)
                {
                    case CommentEvent.EventTypeOneofCase.Created:
                        if (_onCommentCreated != null)
                            await _onCommentCreated.Invoke(ProtoToDto(evt.Created));
                        break;
                    case CommentEvent.EventTypeOneofCase.Updated:
                        if (_onCommentUpdated != null)
                            await _onCommentUpdated.Invoke(ProtoToDto(evt.Updated));
                        break;
                    case CommentEvent.EventTypeOneofCase.DeletedCommentId:
                        if (_onCommentDeleted != null)
                            await _onCommentDeleted.Invoke(new CommentDeletedDto(evt.DeletedCommentId));
                        break;
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (RpcException ex) when (ex.StatusCode == StatusCode.Cancelled) { }
    }

    // ─── Comment methods ──────────────────────────────────────────────────────

    /// <inheritdoc/>
    public async Task<List<CommentDto>> ListOpenCommentsAsync(string documentId, CancellationToken ct = default)
    {
        var response = await _commentsClient.ListOpenCommentsAsync(
            new ListCommentsRequest { DocumentId = documentId }, cancellationToken: ct);
        return response.Comments.Select(ProtoToDto).ToList();
    }

    /// <inheritdoc/>
    public async Task<CommentDto> CreateCommentAsync(string documentId, NewCommentCmd cmd, CancellationToken ct = default)
    {
        var request = new CreateCommentRequest
        {
            PeerId = _peerId,
            DocumentId = documentId,
            Body = cmd.Body,
            AnchorJson = cmd.Anchor is null
                ? string.Empty
                : JsonSerializer.Serialize(cmd.Anchor, JsonOptions),
            ParentCommentId = cmd.ParentCommentId ?? string.Empty
        };
        var response = await _commentsClient.CreateCommentAsync(request, cancellationToken: ct);
        if (!response.Success) throw new InvalidOperationException(response.ErrorMessage);
        return ProtoToDto(response.Comment);
    }

    /// <inheritdoc/>
    public async Task<CommentDto> EditCommentAsync(string documentId, string commentId, string newBody, CancellationToken ct = default)
    {
        var request = new EditCommentRequest
        {
            PeerId = _peerId,
            DocumentId = documentId,
            CommentId = commentId,
            NewBody = newBody
        };
        var response = await _commentsClient.EditCommentAsync(request, cancellationToken: ct);
        if (!response.Success) throw new InvalidOperationException(response.ErrorMessage);
        return ProtoToDto(response.Comment);
    }

    /// <inheritdoc/>
    public async Task<CommentDto> ResolveCommentAsync(string documentId, string commentId, CancellationToken ct = default)
    {
        var request = new CommentActionRequest
        {
            PeerId = _peerId,
            DocumentId = documentId,
            CommentId = commentId
        };
        var response = await _commentsClient.ResolveCommentAsync(request, cancellationToken: ct);
        if (!response.Success) throw new InvalidOperationException(response.ErrorMessage);
        return ProtoToDto(response.Comment);
    }

    /// <inheritdoc/>
    public async Task DeleteCommentAsync(string documentId, string commentId, CancellationToken ct = default)
    {
        var request = new CommentActionRequest
        {
            PeerId = _peerId,
            DocumentId = documentId,
            CommentId = commentId
        };
        var response = await _commentsClient.DeleteCommentAsync(request, cancellationToken: ct);
        if (!response.Success) throw new InvalidOperationException(response.ErrorMessage);
    }

    // ─── Mapping ──────────────────────────────────────────────────────────────

    private static CommentDto ProtoToDto(CommentProto p) => new(
        p.Id,
        p.DocumentId,
        string.IsNullOrEmpty(p.ParentCommentId) ? null : p.ParentCommentId,
        p.AuthorPeerId,
        p.Body,
        string.IsNullOrEmpty(p.AnchorJson)
            ? null
            : JsonSerializer.Deserialize<AnchorDto>(p.AnchorJson, JsonOptions),
        p.AnchoredAtRevision,
        p.CreatedAt.ToDateTimeOffset(),
        p.ResolvedAt?.ToDateTimeOffset(),
        string.IsNullOrEmpty(p.ResolvedByPeerId) ? null : p.ResolvedByPeerId,
        p.IsOrphaned
    );

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        _cts.Cancel();
        try
        {
            await _call.RequestStream.CompleteAsync();
        }
        catch { }
        _call.Dispose();
        _channel.Dispose();
        try { await _listenTask; } catch { }
        if (_commentListenTask != null)
            try { await _commentListenTask; } catch { }
        _cts.Dispose();
    }

    private static AwarenessState FromProto(AwarenessStateProto proto)
    {
        return new AwarenessState(
            proto.PeerId,
            JsonDocument.Parse(proto.DataJson).RootElement.Clone(),
            proto.LastUpdated.ToDateTimeOffset()
        );
    }
}

File: src/OpStream.Client.Transports.gRPC/gRPCOpStreamClient.js
/**
 * A gRPC-based implementation of the OpStream client transport for HTML + JS clients.
 * Handles bidirectional streaming and request-response patterns over a single stream.
 * Note: Bidirectional streaming requires a compatible transport in the browser (e.g., websockets).
 */
export class gRPCOpStreamClient {
    /**
     * Initializes a new instance of the gRPCOpStreamClient.
     * @param {string} address The URL of the gRPC server.
     * @param {Object} options Configuration options to provide the generated classes.
     * @param {any} options.ClientClass The generated OpStreamServicePromiseClient class.
     * @param {any} options.CommentsClientClass The generated OpStreamCommentsServicePromiseClient class.
     * @param {any} options.Messages An object containing the generated protobuf message classes.
     */
    constructor(address, options = {}) {
        const ClientClass = options.ClientClass || (typeof window !== 'undefined' ? window.OpStreamServicePromiseClient : null);
        const CommentsClientClass = options.CommentsClientClass || (typeof window !== 'undefined' ? window.OpStreamCommentsServicePromiseClient : null);
        this.Messages = options.Messages || (typeof window !== 'undefined' ? window : null);

        if (!ClientClass || !CommentsClientClass || !this.Messages) {
            throw new Error("gRPC-Web client classes not found. Ensure they are loaded globally or provided in the options.");
        }

        // Equivalent to Guid.NewGuid()
        this._peerId = typeof crypto !== 'undefined' && crypto.randomUUID 
            ? crypto.randomUUID() 
            : Math.random().toString(36).substring(2, 15);

        this._client = new ClientClass(address);
        this._commentsClient = new CommentsClientClass(address);
        
        this._pendingRequests = new Map();

        // ── Events ────────────────────────────────────────────────────────
        this.onReceiveOp = null;
        this.onDisconnected = null;
        this.onReceiveAwareness = null;
        this.onPeerDisconnected = null;

        this.onCommentCreated = null;
        this.onCommentUpdated = null;
        this.onCommentDeleted = null;

        // Initialize bidirectional stream
        // Note: For standard grpc-web in browser, bidirectional stream requires specific transport (like @improbable-eng/grpc-web websocket)
        if (typeof this._client.connect === 'function') {
            this._call = this._client.connect();
            this._listenLoop();
        } else {
            console.warn("connect() method not found on client. Bidirectional streaming might not be supported by this stub.");
        }
    }

    _listenLoop() {
        if (!this._call || typeof this._call.on !== 'function') return;

        this._call.on('data', async (message) => {
            const correlationId = message.getCorrelationId();
            if (correlationId && this._pendingRequests.has(correlationId)) {
                const pending = this._pendingRequests.get(correlationId);
                this._pendingRequests.delete(correlationId);
                pending.resolve(message);
                return; // Handled as a request-response
            }

            // Handle server push events
            if (message.hasReceiveOpEvent()) {
                if (this.onReceiveOp) {
                    const evt = message.getReceiveOpEvent();
                    await this.onReceiveOp(evt.getPayload(), evt.getNewRevision());
                }
            } else if (message.hasReceiveAwarenessEvent()) {
                if (this.onReceiveAwareness) {
                    const evt = message.getReceiveAwarenessEvent();
                    const states = evt.getAwarenessList().map(p => this._fromAwarenessProto(p));
                    await this.onReceiveAwareness(states);
                }
            } else if (message.hasPeerDisconnectedEvent()) {
                if (this.onPeerDisconnected) {
                    const evt = message.getPeerDisconnectedEvent();
                    this.onPeerDisconnected(evt.getPeerId());
                }
            }
        });

        this._call.on('error', (err) => {
            if (this.onDisconnected) {
                this.onDisconnected(err);
            }
        });

        this._call.on('end', () => {
            if (this.onDisconnected) {
                this.onDisconnected(new Error("Stream ended by server"));
            }
        });
    }

    /**
     * Connects to the document session.
     */
    async connectAndJoinAsync(documentId, documentType) {
        const correlationId = typeof crypto !== 'undefined' && crypto.randomUUID ? crypto.randomUUID() : Math.random().toString(36).substring(2);
        
        return new Promise((resolve, reject) => {
            this._pendingRequests.set(correlationId, { resolve, reject });

            const joinRequest = new this.Messages.JoinRequest();
            joinRequest.setDocumentId(documentId);
            joinRequest.setDocumentType(documentType);
            joinRequest.setClientProtoVersion(1);

            const clientMessage = new this.Messages.ClientMessage();
            clientMessage.setCorrelationId(correlationId);
            clientMessage.setJoinRequest(joinRequest);

            this._call.write(clientMessage);
        }).then((response) => {
            if (!response.hasJoinResponse()) {
                throw new Error("Unexpected response during Join");
            }

            const jr = response.getJoinResponse();
            
            // Start listening to comments for this document
            this._listenCommentsLoop(documentId);

            return {
                revision: jr.getRevision(),
                snapshot: jr.getSnapshot(),
                currentAwareness: jr.getAwarenessList().map(p => this._fromAwarenessProto(p))
            };
        });
    }

    /**
     * Sends an operation payload.
     */
    async sendOpAsync(documentId, payload, baseRevision) {
        const correlationId = typeof crypto !== 'undefined' && crypto.randomUUID ? crypto.randomUUID() : Math.random().toString(36).substring(2);

        return new Promise((resolve, reject) => {
            this._pendingRequests.set(correlationId, { resolve, reject });

            const opRequest = new this.Messages.OpRequest();
            opRequest.setDocumentId(documentId);
            opRequest.setPayload(payload);
            opRequest.setBaseRevision(baseRevision);

            const clientMessage = new this.Messages.ClientMessage();
            clientMessage.setCorrelationId(correlationId);
            clientMessage.setOpRequest(opRequest);

            this._call.write(clientMessage);
        }).then((response) => {
            if (!response.hasOpResponse()) {
                throw new Error("Unexpected response during SendOp");
            }

            const or = response.getOpResponse();
            return {
                success: or.getSuccess(),
                newRevision: or.getNewRevision(),
                errorMessage: or.getErrorMessage()
            };
        });
    }

    /**
     * Sends an awareness update payload.
     */
    async sendAwarenessAsync(documentId, data) {
        const awarenessRequest = new this.Messages.AwarenessRequest();
        awarenessRequest.setDocumentId(documentId);
        // Supports stringified JSON or an object
        awarenessRequest.setDataJson(typeof data === 'string' ? data : JSON.stringify(data));

        const clientMessage = new this.Messages.ClientMessage();
        clientMessage.setCorrelationId(""); // No correlation needed for one-way messages
        clientMessage.setAwarenessRequest(awarenessRequest);

        this._call.write(clientMessage);
    }

    // ─── Comment subscription ─────────────────────────────────────────────────

    _listenCommentsLoop(documentId) {
        if (this._commentCall && typeof this._commentCall.cancel === 'function') {
            this._commentCall.cancel();
        }

        const request = new this.Messages.SubscribeCommentsRequest();
        request.setDocumentId(documentId);

        this._commentCall = this._commentsClient.subscribeComments(request);
        
        if (typeof this._commentCall.on === 'function') {
            this._commentCall.on('data', async (evt) => {
                if (evt.hasCreated() && this.onCommentCreated) {
                    await this.onCommentCreated(this._protoToCommentDto(evt.getCreated()));
                } else if (evt.hasUpdated() && this.onCommentUpdated) {
                    await this.onCommentUpdated(this._protoToCommentDto(evt.getUpdated()));
                } else if (evt.hasDeletedCommentId() && this.onCommentDeleted) {
                    await this.onCommentDeleted({ deletedCommentId: evt.getDeletedCommentId() });
                }
            });

            this._commentCall.on('error', (err) => {
                // Typically cancelled by disposing, ignore
            });
        }
    }

    // ─── Comment methods ──────────────────────────────────────────────────────

    async listOpenCommentsAsync(documentId) {
        const request = new this.Messages.ListCommentsRequest();
        request.setDocumentId(documentId);

        const response = await this._commentsClient.listOpenComments(request);
        return response.getCommentsList().map(p => this._protoToCommentDto(p));
    }

    async createCommentAsync(documentId, cmd) {
        const request = new this.Messages.CreateCommentRequest();
        request.setPeerId(this._peerId);
        request.setDocumentId(documentId);
        request.setBody(cmd.body || cmd.Body);
        
        const anchor = cmd.anchor || cmd.Anchor;
        request.setAnchorJson(anchor ? JSON.stringify(anchor) : "");
        
        const parentId = cmd.parentCommentId || cmd.ParentCommentId;
        request.setParentCommentId(parentId || "");

        const response = await this._commentsClient.createComment(request);
        if (!response.getSuccess()) {
            throw new Error(response.getErrorMessage());
        }
        return this._protoToCommentDto(response.getComment());
    }

    async editCommentAsync(documentId, commentId, newBody) {
        const request = new this.Messages.EditCommentRequest();
        request.setPeerId(this._peerId);
        request.setDocumentId(documentId);
        request.setCommentId(commentId);
        request.setNewBody(newBody);

        const response = await this._commentsClient.editComment(request);
        if (!response.getSuccess()) {
            throw new Error(response.getErrorMessage());
        }
        return this._protoToCommentDto(response.getComment());
    }

    async resolveCommentAsync(documentId, commentId) {
        const request = new this.Messages.CommentActionRequest();
        request.setPeerId(this._peerId);
        request.setDocumentId(documentId);
        request.setCommentId(commentId);

        const response = await this._commentsClient.resolveComment(request);
        if (!response.getSuccess()) {
            throw new Error(response.getErrorMessage());
        }
        return this._protoToCommentDto(response.getComment());
    }

    async deleteCommentAsync(documentId, commentId) {
        const request = new this.Messages.CommentActionRequest();
        request.setPeerId(this._peerId);
        request.setDocumentId(documentId);
        request.setCommentId(commentId);

        const response = await this._commentsClient.deleteComment(request);
        if (!response.getSuccess()) {
            throw new Error(response.getErrorMessage());
        }
    }

    // ─── Mapping ──────────────────────────────────────────────────────────────

    _protoToCommentDto(p) {
        const anchorJson = p.getAnchorJson();
        const resolvedAtProto = p.getResolvedAt();
        
        return {
            id: p.getId(),
            documentId: p.getDocumentId(),
            parentCommentId: p.getParentCommentId() || null,
            authorPeerId: p.getAuthorPeerId(),
            body: p.getBody(),
            anchor: anchorJson ? JSON.parse(anchorJson) : null,
            anchoredAtRevision: p.getAnchoredAtRevision(),
            createdAt: this._toDateTime(p.getCreatedAt()),
            resolvedAt: resolvedAtProto ? this._toDateTime(resolvedAtProto) : null,
            resolvedByPeerId: p.getResolvedByPeerId() || null,
            isOrphaned: p.getIsOrphaned()
        };
    }

    _fromAwarenessProto(proto) {
        return {
            peerId: proto.getPeerId(),
            data: JSON.parse(proto.getDataJson()),
            lastUpdated: this._toDateTime(proto.getLastUpdated())
        };
    }

    _toDateTime(protoTimestamp) {
        if (!protoTimestamp) return null;
        if (typeof protoTimestamp.toDate === 'function') {
            return protoTimestamp.toDate();
        }
        const secs = protoTimestamp.getSeconds ? protoTimestamp.getSeconds() : protoTimestamp.seconds || 0;
        const nanos = protoTimestamp.getNanos ? protoTimestamp.getNanos() : protoTimestamp.nanos || 0;
        return new Date((secs * 1000) + (nanos / 1000000));
    }

    async disposeAsync() {
        if (this._call && typeof this._call.cancel === 'function') {
            this._call.cancel();
        }
        if (this._commentCall && typeof this._commentCall.cancel === 'function') {
            this._commentCall.cancel();
        }
        
        for (const pending of this._pendingRequests.values()) {
            pending.reject(new Error("Client disposed"));
        }
        this._pendingRequests.clear();
        return Promise.resolve();
    }
}
